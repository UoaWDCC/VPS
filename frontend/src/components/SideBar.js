import React, { useContext } from "react";
import { Link, useHistory } from "react-router-dom";
import Button from "@material-ui/core/Button";
import ScenarioContext from "../context/ScenarioContext";
import styles from "../styling/SideBar.module.scss";
import { usePost } from "../hooks/crudHooks";

export default function SideBar() {
  const { currentScenario, setCurrentScenario } = useContext(ScenarioContext);
  const history = useHistory();

  async function createScenario(name = "no name") {
    const newScenario = await usePost(`/api/scenario`, {
      name,
    });
    setCurrentScenario(newScenario);
    // eslint-disable-next-line no-underscore-dangle
    history.push(`/scenario/${newScenario._id}`);
  }

  return (
    <>
      <div className={styles.sideBar}>
        <img
          className={styles.logo}
          src="uoa-med-and-health-sci-logo.png"
          alt="UoA Medical & Health Science Logo"
        />
        <ul className={styles.sideBarList}>
          <li>
            <Button
              className="btn side contained white"
              color="default"
              variant="contained"
              onClick={() => {
                createScenario("default name");
              }}
            >
              Create
            </Button>
          </li>
          <li>
            <Button
              className={`btn side contained white ${
                currentScenario ? "" : styles.buttonDisabled
              }  `}
              color="default"
              variant="contained"
              component={Link}
              to={
                currentScenario
                  ? `/scenario/${currentScenario._id}`
                  : "/scenario/null"
              }
              disabled={!currentScenario}
            >
              Edit
            </Button>
          </li>
        </ul>
      </div>
    </>
  );
}
