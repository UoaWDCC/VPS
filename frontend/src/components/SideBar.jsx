import Button from "@material-ui/core/Button";
import { useContext, useState } from "react";
import { Link, useHistory } from "react-router-dom";
import AuthenticationContext from "../context/AuthenticationContext";
import ScenarioContext from "../context/ScenarioContext";
import AccessLevel from "../enums/route.access.level";
import { useDelete, usePost } from "../hooks/crudHooks";
import styles from "../styling/SideBar.module.scss";
import DeleteButton from "./DeleteButton";
import HelpButton from "./HelpButton";
import CreateScenerioCard from "./CreateScenarioCard";

/**
 * Component used for navigation and executing actions located at the left side of the screen.
 *
 * @component
 * @example
 * return (
 *   <SideBar/ >
 * )
 */
export default function SideBar({ toggleCreateCardVisibility }) {
  const { currentScenario, setCurrentScenario, reFetch } =
    useContext(ScenarioContext);
  const { signOut, getUserIdToken, VpsUser } = useContext(
    AuthenticationContext
  );
  const history = useHistory();

  /** Calls backend end point to create a new empty scenario. */
  async function createScenario(name = "no name") {
    const newScenario = await usePost(
      `/api/scenario`,
      {
        name,
      },
      getUserIdToken
    );
    await usePost(
      `/api/scenario/${newScenario._id}/scene`,
      {
        name: `Scene 0`,
      },
      getUserIdToken
    );
    setCurrentScenario(newScenario);
    // eslint-disable-next-line no-underscore-dangle
    history.push(`/scenario/${newScenario._id}`);
  }

  /** Calls backend end point to switch to the lecturer's dashboard */
  function openDashboard() {
    history.push("/dashboard");
  }

  /** Calls backend end point to delete a scenario. */
  async function deleteScenario() {
    await useDelete(`/api/scenario/${currentScenario._id}`, getUserIdToken);
    setCurrentScenario(null);
    reFetch();
  }

  /** Opens new window to play selected scenario. */
  function playScenario() {
    window.open(`/play/${currentScenario._id}`, "_blank");
  }

  /** Handle creat scenrio card visibility */
  const [isCardVisible, setIsCardVisible] = useState(true);
  function handleCloseCard() {
    setIsCardVisible(false);
  }

  return (
    <>
      {isCardVisible && (
        <CreateScenerioCard
          className="create-scenario-card"
          onClose={handleCloseCard}
        />
      )}
      <div className={styles.sideBar}>
        <img
          draggable="false"
          className={styles.logo}
          src="uoa-logo.png"
          alt="University of Auckland Logo"
        />
        <ul className={styles.sideBarList}>
          <li>
            <Button
              className="btn side contained white"
              color="default"
              variant="contained"
              onClick={() => {
                createScenario("default name");
              }}
            >
              Create
            </Button>
          </li>
          {VpsUser.role === AccessLevel.STAFF ? (
            <li>
              <Button
                className={`btn side contained white ${
                  currentScenario ? "" : "disabled"
                }  `}
                color="default"
                variant="contained"
                onClick={openDashboard}
                disabled={!currentScenario}
              >
                Dashboard
              </Button>
            </li>
          ) : (
            ""
          )}
          <li>
            <Button
              className={`btn side contained white ${
                currentScenario ? "" : "disabled"
              }  `}
              color="default"
              variant="contained"
              onClick={playScenario}
              disabled={!currentScenario}
            >
              Play
            </Button>
          </li>
          <li>
            <Button
              className={`btn side contained white ${
                currentScenario ? "" : "disabled"
              }  `}
              color="default"
              variant="contained"
              component={Link}
              to={
                currentScenario
                  ? `/scenario/${currentScenario._id}`
                  : "/scenario/null"
              }
              disabled={!currentScenario}
            >
              Edit
            </Button>
          </li>
          <li>
            <DeleteButton
              className="btn side contained"
              color="default"
              variant="contained"
              disabled={!currentScenario}
              onClick={deleteScenario}
            >
              Delete
            </DeleteButton>
          </li>
          <li>
            <Button
              className="btn side contained white"
              color="default"
              variant="contained"
              onClick={signOut}
            >
              Logout
            </Button>
          </li>
          <li>
            <HelpButton isSidebar />
          </li>
        </ul>
      </div>
    </>
  );
}
