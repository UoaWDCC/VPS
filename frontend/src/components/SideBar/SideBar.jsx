import { useContext, useState } from "react";
import { useHistory } from "react-router-dom";
import AuthenticationContext from "../../context/AuthenticationContext";
import ScenarioContext from "../../context/ScenarioContext";
import { useDelete, usePost } from "../../hooks/crudHooks";
import CreateScenerioCard from "../CreateScenarioCard/CreateScenarioCard";
import DeleteModal from "../DeleteModal";
import HelpButton from "../HelpButton";
import styles from "./SideBar.module.scss";

import AddCircleOutlineRoundedIcon from "@mui/icons-material/AddCircleOutlineRounded";
import EditRoundedIcon from "@mui/icons-material/EditRounded";
import LogoutIcon from "@mui/icons-material/Logout";
import PlayArrowRoundedIcon from "@mui/icons-material/PlayArrowRounded";

/**
 * Component used for navigation and executing actions located at the left side of the screen.
 *
 * @component
 * @example
 * return (
 *   <SideBar/ >
 * )
 */
export default function SideBar() {
  const { currentScenario, setCurrentScenario, reFetch } =
    useContext(ScenarioContext);
  const { signOut, getUserIdToken } = useContext(AuthenticationContext);
  const history = useHistory();

  /** Calls backend end point to create a new empty scenario. */
  async function createScenario(name = "no name") {
    const newScenario = await usePost(
      `/api/scenario`,
      {
        name,
      },
      getUserIdToken
    );
    await usePost(
      `/api/scenario/${newScenario._id}/scene`,
      {
        name: `Scene 1`,
      },
      getUserIdToken
    );
    setCurrentScenario(newScenario);

    history.push(`/scenario/${newScenario._id}`);
  }

  /** Calls backend end point to delete a scenario. */
  async function deleteScenario() {
    await useDelete(`/api/scenario/${currentScenario._id}`, getUserIdToken);
    setCurrentScenario(null);
    reFetch();
  }

  /** Opens new window to play selected scenario. */
  function playScenario() {
    window.open(`/play/${currentScenario._id}`, "_blank");
  }

  /** Handle creat scenrio card visibility */
  const [isCardVisible, setIsCardVisible] = useState(false);
  function handleCloseCard() {
    setIsCardVisible(false);
  }
  function handleOpenCard() {
    setIsCardVisible(true);
  }

  return (
    <>
      {isCardVisible && (
        <CreateScenerioCard
          className="create-scenario-card"
          onCreate={createScenario}
          onClose={handleCloseCard}
        />
      )}

      {/* Main sidebar */}
      <div className={`${styles.sideBar} bg-uoa-blue `}>
        {/* UoA logo container */}
        <div className="flex-0 p-7">
          <img
            draggable="false"
            className={styles.logo}
            src="uoa-logo.png"
            alt="University of Auckland Logo"
          />
        </div>

        {/* Button containers */}
        <div className="flex flex-col w-full flex-1 justify-between pb-5">
          <ul className={`${styles.sideBarList}`}>
            <li>
              <button
                className="btn vps font-mono"
                onClick={() => {
                  handleOpenCard();
                }}
              >
                <AddCircleOutlineRoundedIcon />
                <span className="min-w-12">Create</span>
              </button>
            </li>
            <li>
              <button
                className="btn vps font-mono"
                onClick={playScenario}
                disabled={!currentScenario}
              >
                <PlayArrowRoundedIcon />
                <span className="min-w-12">Play</span>
              </button>
            </li>
            <li>
              <button
                className="btn vps font-mono"
                disabled={!currentScenario}
                onClick={() => {
                  history.push(`/scenario/${currentScenario._id}`);
                }}
              >
                <EditRoundedIcon />
                <span className="min-w-12">Edit</span>
              </button>
            </li>
            <li>
              {/* Need to add the stuff to check user role */}
              <button
                className="btn vps font-mono"
                disabled={!currentScenario}
                onClick={() => {
                  history.push(`/dashboard/${currentScenario._id}`);
                }}
              >
                <span className="min-w-12">Dashboard</span>
              </button>
            </li>
            <li>
              <DeleteModal
                onDelete={deleteScenario}
                currentScenario={currentScenario}
              />
            </li>
          </ul>

          <ul className={styles.sideBarList}>
            <li>
              <button className="btn vps font-mono" onClick={signOut}>
                <LogoutIcon /> <span className="min-w-13">Logout</span>
              </button>
            </li>
            <li>
              <HelpButton isSidebar />
            </li>
          </ul>
        </div>
      </div>
    </>
  );
}
