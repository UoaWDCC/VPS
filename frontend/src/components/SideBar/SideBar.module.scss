.sideBar {
  min-width: 250px;
  width: 15vw;
  height: 100vh;
  background-color: #035084;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  overflow-y: auto; 
  max-height: 100vh;
  scrollbar-width: none; 

  .logo {
    margin: 10px;
    width: 230px;
  }

  .sideBarList {
    list-style: none;
    padding: 0;

    li:not(:last-child) {
      margin-bottom: 10%;
    }

    .buttonDisabled {
      opacity: 0.3;
    }
  }

  .sideBarText {
    color: white;
    font-size: 1.5rem;
    margin: 0;
    padding: 2.5em 0;
  }

  .googleButton {
    display: flex;
    flex-direction: row;
    align-items: center;
    gap: 0.5em;
    background-color: #f4f4f4;
    color: #035084;
    font-size: 1rem;
    padding: 0.75em 1em;
    border: none;
    border-radius: 5px;
    cursor: pointer;
  }

  .sponsorText {
    color: white;
    font-size: 1rem;
    margin: 0;
    padding: 15rem 4em;
  }
}
