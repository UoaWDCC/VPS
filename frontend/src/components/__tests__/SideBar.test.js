import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import SideBar from "../SideBar";

test("Side Bar component snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <SideBar />
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
