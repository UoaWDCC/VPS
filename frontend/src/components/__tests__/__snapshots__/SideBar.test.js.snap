// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Side Bar component snapshot test 1`] = `
<div
  className="sideBar"
>
  <img
    alt="University of Auckland Logo"
    className="logo"
    draggable="false"
    src="uoa-logo.png"
  />
  <ul
    className="sideBarList"
  >
    <li>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained btn side contained white"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Create
        </span>
      </button>
    </li>
    
    <li>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained btn side contained white disabled   Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Play
        </span>
      </button>
    </li>
    <li>
      <a
        aria-disabled={true}
        className="MuiButtonBase-root MuiButton-root MuiButton-contained btn side contained white disabled   Mui-disabled Mui-disabled"
        href="/scenario/null"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <span
          className="MuiButton-label"
        >
          Edit
        </span>
      </a>
    </li>
    <li>
      <button
        className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-1 MuiButton-contained btn side contained Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Delete
        </span>
      </button>
    </li>
    <li>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained btn side contained white"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Logout
        </span>
      </button>
    </li>
    <li>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined btn outlined white side"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
              />
            </svg>
          </span>
          Help
        </span>
      </button>
    </li>
  </ul>
</div>
`;
